packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts

```diff
@@ -92,14 +104,72 @@ export class SnsPublish extends sfn.TaskStateBase {
    * @internal
    */
   protected _renderTask(): any {
+
     return {
       Resource: integrationResourceArn('sns', 'publish', this.integrationPattern),
       Parameters: sfn.FieldUtils.renderObject({
         TopicArn: this.props.topic.topicArn,
         Message: this.props.message.value,
         MessageStructure: this.props.messagePerSubscriptionType ? 'json' : undefined,
+        MessageAttributes: renderMessageAttributes(this.props.messageAttributes),
         Subject: this.props.subject,
       }),
     };
   }
 }
+
+function renderMessageAttributes(attributes?: { [key: string]: any }): any {
+  if (attributes === undefined) { return undefined; }
+  const attrs: { [key: string]: messageAttributeValue } = {};
+  for (const name of Object.keys(attributes)) {
+    const value = handleMessageAttributeValue(attributes[name]);
+    attrs[name] = value;
+  };
+  return sfn.TaskInput.fromObject(attrs).value;
+}
+
+interface messageAttributeValue {
+  DataType: string;
+  StringValue?: string;
+  BinaryValue?: string;
+};
+
+const STRING = 'String';
+const STRING_ARRAY = 'String.Array';
+const NUMBER = 'Number';
+const BINARY = 'Binary';
+
+function handleMessageAttributeValue(attributeValue: any): messageAttributeValue {
+  if (attributeValue instanceof sfn.TaskInput) {
+    return { DataType: STRING, StringValue: attributeValue.value };
+  }
+  if (isByteArray(attributeValue)) {
+    return { DataType: BINARY, BinaryValue: '' };
+  }
+  if (Array.isArray(attributeValue)) {
+    // validates the primitives
+    attributeValue.forEach(v => handlePrimitiveValues(v));
```

Comment:
I think we may get into trouble here at some point if `handlePrimitiveValues` diverges from the spec of what we can put in this array (if we can make tokens work then we should be able to allow a token in this array). Let's create a separate validation function